Add ArtifcactHUB pre-release annotation to the Helm chart

wallrj opened the pull request:
In https://github.com/jetstack/jetstack-charts/pull/15#pullrequestreview-648213790 @wallrj wrote:
> think it'd be nice to instead use the Charts.yaml annotations artifacthub.io/prerelease , so that people get to see the alphas and betas on Artifcact Hub
> 
> https://artifacthub.io/docs/topics/annotations/helm/

I hacked away at Bazel + shell script to parse the `//:version` artifact for `\d.\d.\d` to see if it is a release or a prerelease and use that to fill in in the `artifacthub.io/prerelease` annotation.

There may be a cleaner way to do this.

We're aiming to make it so that the next alpha / beta release of cert-manager doesn't show up here:
 * https://artifacthub.io/packages/helm/cert-manager/cert-manager

And when 1.4.0 is finally released, that will show up.

**Release note**:
```release-note
NONE
```

/kind cleanup

wallrj commented:
It worked on my laptop:

```
Target //deploy/charts/cert-manager:cert-manager up-to-date:
  bazel-bin/deploy/charts/cert-manager/cert-manager.tgz
INFO: Elapsed time: 1.333s, Critical Path: 0.76s
INFO: 7 processes: 1 internal, 6 linux-sandbox.
INFO: Build completed successfully, 7 total actions
[richard@drax cert-manager]$ helm show chart ./bazel-bin/deploy/charts/cert-manager/cert-manager.tgz 
annotations:
  artifacthub.io/prerelease: "true"
apiVersion: v1
appVersion: v1.4.0-alpha.1.78-15e9043dab35ec
description: A Helm chart for cert-manager
home: https://github.com/jetstack/cert-manager
icon: https://raw.githubusercontent.com/jetstack/cert-manager/master/logo/logo.png
keywords:
- cert-manager
- kube-lego
- letsencrypt
- tls
maintainers:
- email: cert-manager-maintainers@googlegroups.com
  name: cert-manager-maintainers
name: cert-manager
sources:
- https://github.com/jetstack/cert-manager
version: v1.4.0-alpha.1.78-15e9043dab35ec

```

But not in CI, due to missing envsubst. :shrug:

irbekrm commented:
> But not in CI, due to missing envsubst

Maybe we can somehow pass the PRELEASE var as an argument to `helm template` command [here-ish](https://github.com/jetstack/cert-manager/blob/master/build/helm.bzl#L106-L113) (I've only looked at it very briefly so not 100% sure)

wallrj commented:
> > But not in CI, due to missing envsubst
> 
> Maybe we can somehow pass the PRELEASE var as an argument to `helm template` command [here-ish](https://github.com/jetstack/cert-manager/blob/master/build/helm.bzl#L106-L113) (I've only looked at it very briefly so not 100% sure)

No, because `helm template` is used for generating the static manifest yamls. What we need to do is modify the Chart.yaml file before it gets packaged up in the chart archive. I looked at the `helm package --help` options, but there are no options for modifying the `Chart.yaml` file there.

SgtCoDFish commented:
My limited experience with `awk` aside, this looks good.

Did you consider a regex like `(alpha|beta)`?

`^v[0-9]+.[0-9]+.[0-9]+$` works but could incorrectly mark, say, `v1.4.0.1`, as a pre-release if we did a really minor patch release (although I recognise that that deviates from our current versioning scheme). `(alpha|beta)` wouldn't match `v1.4.0.1`, as an example.

if we're going with it as-is, should it be `^v[0-9]+\.[0-9]+\.[0-9]+$`? otherwise with `.` it could match `v1a4b0` (awk syntax may be different here, I don't know)

maelvls commented on build/helm.bzl:
```diff
@@ -121,3 +121,28 @@ def helm_tmpl(
+        stamp = 1,
+        outs = ["Chart.yaml"],
+        cmd = """
+        awk '
+          BEGIN{
+            getline v < "$(location %s)"
+            pr = match(v, "^v[0-9]+.[0-9]+.[0-9]+$$") == 0 ? "true" : "false"
+          }
+          /{PRERELEASE}/{
+            gsub("{PRERELEASE}", pr)
+          }1
+        ' $(location %s) > $@
+        """ % (version_file, chart_yaml_template),
```
For example:

```suggestion
        # We don't want to see our alphas to show on ArtifactHub as releases.
        # To prevent pre-releases from appearing, we use the ArtifactHub-specific
        # annotation: 
        #   artifacthub.io/prerelease: "true"
        awk '
          BEGIN{
            getline v < "$(location %s)"
            pr = match(v, "^v[0-9]+.[0-9]+.[0-9]+$$") == 0 ? "true" : "false"
          }
          /{PRERELEASE}/{
            gsub("{PRERELEASE}", pr)
          }1
        ' $(location %s) > $@
        """ % (version_file, chart_yaml_template),
```

maelvls reviewed (commented):
One suggestion, but that looks good to me otherwise!!

/hold
/lgtm

wallrj commented:
> Did you consider a regex like (alpha|beta)?

I did consider that, but it would not catch other suffixes, like `-pre` or `-dirty`. 
Since I don't know what pre-release suffixes there might be in the future, I thought it'd be safer to use a very strict regexp for the non-prerelease case.

> 
> ^v[0-9]+.[0-9]+.[0-9]+$ works but could incorrectly mark, say, v1.4.0.1, as a pre-release if we did a really minor patch release (although I recognise that that deviates from our current versioning scheme). (alpha|beta) wouldn't match v1.4.0.1, as an example.

I don't think that example is a valid semver, so we shouldn't ever get that.
More likely that we introduce some new pre-release suffix such as `-rc.1` for release candidates etc, which is why I chose to match on the release version rather than on pre-release versions.

> if we're going with it as-is, should it be ^v[0-9]+\.[0-9]+\.[0-9]+$? otherwise with . it could match v1a4b0 (awk syntax may be different here, I don't know)

Yeah, I kept seeing the following warning when I escaped the `.`:
```
INFO: From Executing genrule //deploy/charts/cert-manager:annotated_chart_yaml:
awk: cmd. line:4: warning: escape sequence `\.' treated as plain `.'

```

But I've now realized that I should delimit the regexps with forward slashes rather than quotes `/[0-9]+\. .../` and that allows me to escape the dot.

wallrj commented:
@SgtCoDFish and @maelvls 

I addressed your comments.

You can test this locally as follows:


```sh
$ bazel build //deploy/charts/cert-manager:cert-manager 
INFO: Analyzed target //deploy/charts/cert-manager:cert-manager (1 packages loaded, 8 targets configured).
INFO: Found 1 target...
Target //deploy/charts/cert-manager:cert-manager up-to-date:
  bazel-bin/deploy/charts/cert-manager/cert-manager.tgz
INFO: Elapsed time: 1.154s, Critical Path: 0.65s
INFO: 7 processes: 1 internal, 6 linux-sandbox.
INFO: Build completed successfully, 7 total actions
[richard@drax cert-manager]$ helm show chart ./bazel-bin/deploy/charts/cert-manager/cert-manager.tgz 
annotations:
  artifacthub.io/prerelease: "true"
apiVersion: v1
appVersion: v1.4.0-alpha.1.80-03a794f94c8984

...
```

Create a local tag and repeat:

```sh
$ git tag v1.4.0

$ bazel build //deploy/charts/cert-manager:cert-manager 
INFO: Analyzed target //deploy/charts/cert-manager:cert-manager (1 packages loaded, 8 targets configured).
INFO: Found 1 target...
Target //deploy/charts/cert-manager:cert-manager up-to-date:
  bazel-bin/deploy/charts/cert-manager/cert-manager.tgz
INFO: Elapsed time: 1.170s, Critical Path: 0.65s
INFO: 7 processes: 1 internal, 6 linux-sandbox.
INFO: Build completed successfully, 7 total actions

$ helm show chart ./bazel-bin/deploy/charts/cert-manager/cert-manager.tgz 
annotations:
  artifacthub.io/prerelease: "false"
apiVersion: v1
appVersion: v1.4.0
...

$ git tag v1.4.0 --delete
```

wallrj commented:
> Is $(location %s) bazel-specific?

Yes, Bazel expands those to have the build time path to the supplied bazel target.
E.g. `//:version` becomes `bazel-bin/version`

```sh
$ cat bazel-bin/version 
v1.4.0-alpha.1.80-03a794f94c8984

```

wallrj commented on build/helm.bzl:
```diff
@@ -121,3 +121,36 @@ def helm_tmpl(
+          }
+          /{PRERELEASE}/{
+            gsub(/{PRERELEASE}/, pr)
+          }1
```
I don't know what this trailing `1` does. I copied it from https://unix.stackexchange.com/a/141398
Without it, the script fails :shrug:

wallrj replied:
Ah https://stackoverflow.com/a/20263611 explains it, but I'll use `{print}` instead.

maelvls commented:
Well done!!
/lgtm

SgtCoDFish commented:
/unhold

I also like this - I'm a little nervous about the `awk` just because it's hard to read (which isn't a criticism of this code specifically but rather the language and the context in which it's being used) but if the longer term goal is to replace bazel then we can easily use `yq` or something at some point down the road.
